### ISSUE TITLE
Incorrect Unsat Core with smt.mbqi=false

### ISSUE THREAD

**Issue by Heizmann** [Z3Prover/z3#158]
While executing z3 eeef4d29d68285a505468715bccc892db76f1267 with the argument **auto_config=false smt.mbqi=false** I obtain the unsat core
`(ssa_3_conjunct10 ssa_1_conjunct19 ssa_1_conjunct13 ssa_3_conjunct20 ssa_5_conjunct0 ssa_3_conjunct16 ssa_3_conjunct9 ssa_3_conjunct19 ssa_1_conjunct0 ssa_0_conjunct1 ssa_0_conjunct0 ssa_1_conjunct20 ssa_3_conjunct15 ssa_6_conjunct0)`
on the following SMT script.
http://www.informatik.uni-freiburg.de/~heizmann/20150711-NoMbqi-UnsatCore-Bug.smt2.zip
This unsat core is not correct. I demonstrated this at the end of the script. I asserted only the conjuncts from the alleged unsat core, but the result of a *check-sat* is sat.

If you delete all apparently superfluous SMT commands before the *get-unsat-core*, the bug is not reproduceable.

**Comment by NikolajBjorner** [comment]
The current unstable branch produces a different result at this point.

**Comment by NikolajBjorner** [comment]
I also added a built-in core validation option. It can be enabled using the option smt.core.validate=true. It throws an exception if the core is not verified separately. At this point the repro file goes through and the cores are validated.

**Comment by NikolajBjorner** [comment]
The bug is currently not reproducible.

**Comment by Heizmann** [comment]
Using ef7915858bca5e759ee0a34df2ae976e7e507bef (the latest version I obtained an hour ago) I am also not able to reproduce the bug.
Some notes about this bug:
We found it by applying [Ultimate Automizer](http://ultimate.informatik.uni-freiburg.de/automizer/ Ultimate Automizer) to hundreds of [SV-COMP benchmarks](http://sv-comp.sosy-lab.org) in different settings. To the best of our knowledge the tool returned only on one C file an incorrect result because of this bug.
On the original several GiB large SMT script (which I don't have any more) the incorrect unsat core occured nondetermnistically. Probably this was already a bug, because I guess Z3 is supposed to be determnistic. Only after reducing the input to our tool and the SMT script I was able to reproduce the incorrect unsat core deterministically.
After finding the bug, we used Z3 only with smt.mbqi. In the next weeks I might run again some benchmarks where our tools uses Z3 also without smt.mbqi.

**Comment by NikolajBjorner** [comment]
The bug should be independent of smt.mbqi. Try to use "smt.core.validate=true" as well (use your own validation and if it fails check if smt.core.validate fails, or check if smt.core.validate fails in the first hand).

**Comment by wintersteiger** [comment]
It's been quiet here for a while; can we consider this issue as resolved, or is there more testing to be done?

**Comment by wintersteiger** [comment]
Closing this issue no more incorrect results have been reported. If any problems remain please submit a new issue or ask us to re-open this one.
